smithy-utils/src/main/java/software/amazon/smithy/utils/DefaultBuilderRef.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.utils;
+
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Default implementation of {@link BuilderRef}.
+ *
+ * @param <T> The type of value being built, borrowed, and copied.
+ */
+final class DefaultBuilderRef<T> implements BuilderRef<T> {
```

Comment:
should this be documented as explicitly not threadsafe?

Reply:
Done